src/chronify/sqlalchemy/functions.py

```diff
@@ -1,24 +1,43 @@
-from typing import Literal, TypeAlias
+"""This file provides functions to read and write the database as efficiently as possible.
+The default behavior of sqlalchemy is to convert data into rows of tuples in Python, which
+is very slow. This code attempts to bypass Python as much as possible through Arrow tables
+in memory.
+"""
+
+from typing import Any, Literal, TypeAlias
+
 import pandas as pd
 import polars as pl
 from numpy.dtypes import ObjectDType
 from pandas import DatetimeTZDtype
 from sqlalchemy import Connection, Selectable
 
+from chronify.exceptions import InvalidOperation
 from chronify.time_configs import DatetimeRange, TimeBaseModel
 
 # Copied from Polars, which doesn't export the type.
 DbWriteMode: TypeAlias = Literal["replace", "append", "fail"]
 
 
 def read_database(
-    query: Selectable | str, conn: Connection, config: TimeBaseModel
+    query: Selectable | str, conn: Connection, config: TimeBaseModel, params: Any = None
 ) -> pd.DataFrame:
     """Read a database query into a Pandas DataFrame."""
-    df = pl.read_database(query, connection=conn).to_pandas()
-    if isinstance(config, DatetimeRange):
-        df = _convert_database_output_for_datetime(df, conn, config)
-    return df
+    if conn.engine.name == "duckdb":
+        if isinstance(query, str):
+            df = conn._dbapi_connection.driver_connection.sql(query, params=params).to_df()  # type: ignore
+        else:
+            df = conn.execute(query).cursor.fetch_df()  # type: ignore
+    elif conn.engine.name == "sqlite":
+        df = pd.read_sql(query, conn, params=params)
+        if isinstance(config, DatetimeRange):
+            _convert_database_output_for_datetime(df, config)
```

Comment:
@lixiliu You'll notice that this modifies the dataframe we just read from the db. This should be safe and avoids unnecessary copies. Please raise any specific concerns.